geth/peers/peerpool.go

```diff
@@ -152,18 +152,27 @@ func (p *PeerPool) handleServerPeers(server *p2p.Server, events <-chan *p2p.Peer
 					retryDiscv5 = time.After(0)
 				}
 			case p2p.PeerEventTypeAdd:
-				log.Debug("confirm peer added", "ID", event.Peer)
-				if p.stopOnMax && p.handleAddedPeer(server, event.Peer) {
-					log.Debug("closing discv5 connection", "server", server.Self())
-					server.DiscV5.Close()
-					server.DiscV5 = nil
-					p.feed.Send(Discv5Closed)
-				}
+				p.handleAddedEvent(server, event)
 			}
 		}
 	}
 }
 
+// handledAddedEvent maintains logic of stopping discovery server and added mainly
+// for easier testing.
+func (p *PeerPool) handleAddedEvent(server *p2p.Server, event *p2p.PeerEvent) {
+	log.Debug("confirm peer added", "ID", event.Peer)
+	if p.stopOnMax && p.handleAddedPeer(server, event.Peer) {
+		if server.DiscV5 == nil {
+			return
+		}
+		log.Debug("closing discv5 connection", "server", server.Self())
+		server.DiscV5.Close()
```

Comment:
How do we make sure that between the check above and this call, `DiscV5` was not shut down? Is this the only place when it can be shut down?

Reply:
during app runtime this must be the only place, but when app will shutdown we will ensure that this loop is exited before stopping geth node